### ISSUE TITLE
EE10 `ServletChannelState` input FSM should be reviewed

### ISSUE THREAD

**Issue by lorban** [eclipse/jetty.project#10357]
**Jetty version(s)**
12.0.0

**Jetty Environment**
EE10

**Description**
While fixing https://github.com/eclipse/jetty.project/issues/10315, it became apparent that the implementation behind the FSM backing `HttpInput` should be reviewed entirely.

The impedance mismatch between the Jetty 10/11 `needContent()` / `produceContent()` contract vs the Jetty 12 `demand()` / `read()` one means the state checks and transitions need to be placed slightly differently in certain situations.

For instance, #10315 was about the fact that `needContent()` fills, parses and registers for fill interest if necessary while being idempotent; doing the same thing in EE10 requires calling `read()` then `demand()` if necessary, but also needs to take into account that `read()` must not be called while there is a pending demand as the idempotency is gone from the EE10-backing API. Fixing that bug required checking that when using the blocking API, `ServletInputStream.isReady()` must not call `read()` while the FSM is in `UNREADY` state otherwise we would call `read()` during a pending demand. Fixing the bug required changing the state transitions so that `BlockingContentProducer` moves back to `IDLE` state when a demand has been serviced. This means the comments in `BlockingContentProducer.onContentProducible()` which describe this mechanism don't apply anymore to EE10 (`HttpChannelState.asyncError()` went away for instance) and the references to the tests covering this mechanism are stale too.

Another point requiring investigation is that some tests like `ProxyServletTest.testExpect100ContinueRespond100ContinueSomeRequestContentThenFailure` log the following warning when executed, while passing and being apparently correct:
```
java.lang.IllegalStateException: s=HANDLING rs=BLOCKING os=OPEN is=IDLE awp=false se=false i=false al=0
	at org.eclipse.jetty.ee10.servlet.ServletChannelState.handling(ServletChannelState.java:357)
	at org.eclipse.jetty.ee10.servlet.ServletChannel.handle(ServletChannel.java:462)
	at org.eclipse.jetty.server/org.eclipse.jetty.server.handler.ContextHandler$ScopedContext.run(ContextHandler.java:1264)
	at org.eclipse.jetty.server/org.eclipse.jetty.server.handler.ContextHandler$ScopedContext.run(ContextHandler.java:1251)
	at org.eclipse.jetty.server/org.eclipse.jetty.server.handler.ContextHandler$ScopedContext.lambda$execute$0(ContextHandler.java:1276)
	at org.eclipse.jetty.util/org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:969)
	at org.eclipse.jetty.util/org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.doRunJob(QueuedThreadPool.java:1194)
	at org.eclipse.jetty.util/org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1149)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

All the state transitions both in blocking and unblocking modes, as well as test coverage (including using H2 / H3) should be re-assessed for EE10 as we know from experience that this FSM is very delicate.

**Comment by gregw** [comment]
@lorban can you start by creating some more tests? Specifically, can you recreate the ISE in a non-proxy test?

**Comment by lorban** [comment]
It appears the `IllegalStateException` isn't related to the FSM in `HttpInput` but to the FSM in `ServletChannel`.

in `handle()` when the state is `SEND_ERROR` we have this code:

```java
AtomicBoolean asyncCompletion = new AtomicBoolean(false);
Callback errorCallback = Callback.from(() ->
{
    if (!asyncCompletion.compareAndSet(false, true))
        _state.scheduleDispatch();
});
errorHandler.handle(getServletContextRequest(), getServletContextResponse(), errorCallback);
if (asyncCompletion.compareAndSet(false, true))
    return;
```

so the `errorHandler` executes, and the code uses an atomic boolean to figure out if the callback is executed before `handle()` returns or not.

The problem is that if the callback does not execute before `errorHandler.handle()` returns, `_state.scheduleDispatch()` submits a job that may execute `ServletChannel.handle()` while the thread that executed `errorHandler.handle()` hasn't yet executed `_state.unhandle()`.  This is why we can see the state still being `HANDLING` during a call to `handle()`.

And since this code is racy, the error isn't always reported and we mistakenly believed the problem went away with the fix for https://github.com/jetty/jetty.project/issues/10315.

**Comment by sbordet** [comment]
Fixed by #10930.
